### DISCUSSION TITLE
USB midi keyboard is not supported

### DISCUSSION THREAD

**Discussion by ilyasni** [probonopd/MiniDexed#258] [Q&A]
The usb keyboard I have is not supported. What do you advise to do? Maybe it's worth adding something to cmdline.txt?
![IMG_3763](https://user-images.githubusercontent.com/42122151/170471386-4508c4bd-29f7-4e59-a1fa-1e999318345b.jpeg)

**Comment by luisfcorreia** [comment]
Hi, 

apparently that device requires drivers as it's not really a USB MIDI Class Compliant Device.

There is nothing MiniDexed can do in order to support it directly.

**Reply by craigyjp** [reply]
Agreed, if you can plug a usb device directly into a Mac or PC without drivers then it's class compliant and will likely work with this project, if your usb device requires a driver then you are pretty much on your own.

**Reply by probonopd** [reply]
@ilyasni which exact model do you have? The Novation Launchkey Mini and Novation Launchkey Mini mk II seem to be advertised as USB class compliant. I'd be somewhat surprised if the MK3 wasn't anymore.

Did you try to remove `usbspeed=full` from `cmdline.txt`?

**Reply by pik33** [reply]
The same thing with Novation Impulse 61. Removing usbspeed=full didn't help. It only enabled the RPi keyboard which didn't work with usb=full.

Edit: And they say "Impulse is class-compliant and USB bus-powered, so you can simply plug it into Mac or Windows and get started straight away, without the need for a power supply."

**Reply by diyelectromusic** [reply]
Well, int1-1-0 is Audio class, Audio Control subclass and int1-3-0 is Audio class, MIDI Streaming subclass, so I wonder what it didn't like in the descriptor...

Presumably it works fine in a Raspbian or other setup?  It might be interesting to see what lsusb says about it...

**Reply by diyelectromusic** [reply]
Ok, so I was looking through the Linux source and there are a whole load of "quirks" for Yamaha devices(!) but not many listed for this device (which shares the same Vendor ID as Novation).  See: https://github.com/torvalds/linux/blob/master/sound/usb/quirks-table.h

... but it also highlights just how quirky, "MIDI compliant" devices actually can be!

As I say, it would be interesting to know what lsusb on Rasbian thinks about it.

**Comment by diyelectromusic** [comment]
Ok, well mapping through that hex dump, you can see a range of entries of the form <size> <descriptor type> <data> that I think map out as follows:
09 02 63 00 02 01 00 80 32 - Configuration descriptor with 2 interfaces
09 04 00 00 00 01 01 00 03 - Interface descriptor with no endpoints - Class 1 (audio), Subclass 1 (control)
09 24 01 00 01 09 00 01 01 - Class specific audio interface descriptor 01 (header)
09 04 01 00 02 01 03 00 03 - Interface descriptor with 2 endpoints - Class 1 (audio), Subclass 3 (MIDI streaming)
07 24 01 00 01 3F 00           - Class specific audio interface descriptor 01 (header)
06 24 02 01 01 04                - Class specific audio interface descriptor 02 (input terminal)
06 24 02 01 04 05                - Class specific audio interface descriptor 02 (input terminal)
09 24 03 01 02 01 01 01 04 - Class specific audio interface descriptor 03 (output terminal)
09 24 03 01 03 01 02 02 05 - Class specific audio interface descriptor 03 (output terminal)
07 05 81 03 40 00 01           - Endpoint descriptor: EP 1 (input)
06 25 01 02 02 03                - Class specific audio endpoint descriptor 01 (general)
07 05 81 03 40 00 01           - Endpoint descriptor: EP 1 (input)
06 25 01 02 02 03                - Class specific audio endpoint descriptor 01 (general)
07 05 02 03 40 00 01           - Endpoint descriptor: EP 2 (output)
06 25 01 02 01 04                - Class specific audio endpoint descriptor 01 (general)
(the rest is after the end of the 0x63 total bytes in the descriptor).

It is complaining about position 0x56 which I think is the start of the third end point descriptor - the second from last line...  but seeing as the first interface says it has no descriptors, but then we end up with three be described, I wonder if that is part of the problem...?

It /looks/ like the circle MIDI device might drop out if it already has an endpoint when it comes to processing the next one, but I don't really know. (see https://github.com/rsta2/circle/blob/master/lib/usb/usbmidi.cpp).  At this point I'd be adding some extra logging in circle to see what is going on, but as I don't have the device I can't.

Anyone else want to comment on the decoding?  Do you agree?  might need to ask @rsta2 about this one ... :)

Kevin

**Reply by diyelectromusic** [reply]
So, turns out my Novation Launchpad Mini is also not supported.  The descriptors play out as follows:
09 02 52 00 02 01 00 80 FA - Config descriptor with 2 interfaces (total size 0052)
09 04 00 00 00 01 01 00 03 - Interface with 0 endpoints Audio class; Control subclass
09 24 01 00 01 09 00 01 01 - Class specific audio interface header
09 04 01 00 02 01 03 00 03 - Interface with 2 endpoints Audio class; MIDI streaming subclass
07 24 01 00 01 2E 00 - Class specific audio interface header
06 24 02 01 01 00 - Class specific audio interface input terminal
09 24 03 01 02 01 01 01 00 - Class specific audio interface output terminal
07 05 81 03 40 00 01 - Endpoint descriptor 1 (input)
05 25 01 01 02 - Class specific audio endpoint descriptor
07 05 02 03 40 00 01 - Endpoint descriptor 2 (output)
05 25 01 01 01 - Class specific audio endpoint descriptor

Once again it fails at the start of the last (output) endpoint descriptor for some reason.  So ignore the "two endpoints" bit from earlier, but at least I might have something I can test against... I'll report back if I get anywhere (unless someone gets there first).

Kevin

**Comment by diyelectromusic** [comment]
Ah, I think it is because the two endpoints are specified with the attributes: 0x03 which means INTERRUPT transfers, but Circle only looks for 0x02 for BULK transfers...

If I change circle to accept INTERRUPT endpoints too, it seems to work... but I have no idea what the implications of that would be!  I'll register an issue with circle...

Update: Issue logged here: https://github.com/rsta2/circle/issues/316

**Comment by probonopd** [comment]
Please try this build:
https://github.com/probonopd/MiniDexed/suites/8451024227/artifacts/374863366

References
* https://github.com/rsta2/circle/issues/316#issuecomment-1256964271
* https://github.com/probonopd/MiniDexed/commit/d53aeb40f505101eb7545fa6ac925d6a0b4bc038

**Reply by diyelectromusic** [reply]
Right, I've just tested this with all the USB Midi controllers I have or can borrow, and the following all seem to work:
* Korg microKEY Air
* Arturia MiniLab MK II
* M-Audio KeyStudio
* CME XKey
* Novation Launchpad Mini
* Raspberry Pi Pico running Circuitpython
* Arduino Pro Micro

All of these apart from the Novation Lauchpad Mini work with the original too.

**Reply by probonopd** [reply]
So you are saying the Novation Lauchpad Mini just started working? Awesome :+1: thanks a lot for helping to make this happen.

**Reply by diyelectromusic** [reply]
Yep. Tbh, it worked with my hacked-together version of Circle, so I thought we had a pretty good chance!

Essentially, USB MIDI INTERRUPT endpoints are designed (as far as I can see) to tell a host that it is supposed to service the USB MIDI within a certain timed response... So, if Circle says it will accept BULK or INTERRUPT, but doesn't quite make the timing constraint, I don't see how it is any worse than what it would have been for a BULK endpoint anyway... so accepting both and then just doing what it was doing anyway is a lot better than not accepting INTERRUPT endpoints at all!

I'd like to know if it now works for the Novation Impulse now.  I think it should.  Someone near to me is selling an Impulse 25 for £50, but I can't justify buying any more MIDI controllers just for testing! :)

Hopefully from a USB MIDI class compliance point of view, this is a major gap now filled!  Until we find the next one :)

Anyway, I'd be really interested to see how people get on now.  Maybe we should start a new Wiki page for supported USB controllers....

(I'd still like to know why that Yamaha wasn't doing anything... and I myself have an unsupported Yamaha digital piano... but that is a job for another day!!)

**Reply by diyelectromusic** [reply]
@pik33 are you able to try this new version with your Novation Impulse 61?  I think there is a good possibility it might work if it is anything like my Novation Launchpad Mini...

Kevin

**Reply by pik33** [reply]
Yes, it now works. Attached the keyboard info from lsusb.
[impulse.txt](https://github.com/probonopd/MiniDexed/files/9640443/impulse.txt)
